# Homework Assignment - Lesson 7: String Manipulation and Date/Time Data

**Student Name:** Mahreen Maknojia

**Student ID:** nsa719

**Date Submitted:** 10/11/2025

**Due Date:** 10/12/2025

---

## Objective

Master string manipulation with `stringr` and date/time operations with `lubridate` for real-world business data cleaning and analysis.

## Learning Goals

By completing this assignment, you will:
- Clean and standardize messy text data using `stringr` functions
- Parse and manipulate dates using `lubridate` functions
- Extract information from text and dates for business insights
- Combine string and date operations for customer segmentation
- Create business-ready reports from raw data

## Instructions

- Complete all tasks in this notebook
- Write your code in the designated TODO sections
- Use the pipe operator (`%>%`) wherever possible
- Add comments explaining your logic
- Run all cells to verify your code works
- Answer all reflection questions

## Datasets

You will work with three CSV files:
- `customer_feedback.csv` - Customer reviews with messy text
- `transaction_log.csv` - Transaction records with dates
- `product_catalog.csv` - Product descriptions needing standardization

---

## Part 1: Data Import and Initial Exploration

**Business Context:** Before cleaning data, you must understand its structure and quality issues.

**Your Tasks:**
1. Load required packages (`tidyverse` and `lubridate`)
2. Import all three CSV files from the `data/` directory
3. Examine the structure and identify data quality issues
4. Display sample rows to understand the data

In [50]:
# Task 1.1: Load Required Packages
# TODO: Load tidyverse (includes stringr)
library(tidyverse)


# TODO: Load lubridate
library(lubridate)


cat("✅ Packages loaded successfully!\n")

✅ Packages loaded successfully!


In [51]:
# Task 1.2: Import Datasets
# Set working directory # Change to your actual path

# TODO: Import customer_feedback.csv into a variable called 'feedback'
feedback <- read_csv("/workspaces/1MS3083/data/customer_feedback.csv")
# TODO: Import transaction_log.csv into a variable called 'transactions'
transactions <- read.csv("/workspaces/1MS3083/data/transaction_log.csv")
# TODO: Import product_catalog.csv into a variable called 'products'
products <- read_csv("/workspaces/1MS3083/data/product_catalog.csv")

cat("✅ Data imported successfully!\n")
cat("Feedback rows:", nrow(feedback), "\n")
cat("Transaction rows:", nrow(transactions), "\n")
cat("Product rows:", nrow(products), "\n")

Rows: 100 Columns: 6
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (3): Customer_Name, Feedback_Text, Contact_Info
dbl  (2): FeedbackID, CustomerID
date (1): Feedback_Date

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 75 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (3): Product_Description, Category, In_Stock
dbl (2): ProductID, Price

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


✅ Data imported successfully!
Feedback rows: 100 
Transaction rows: 150 
Product rows: 75 


In [52]:
# Task 1.3: Initial Data Exploration

cat("=== CUSTOMER FEEDBACK DATA ===\n")
# TODO: Display structure of feedback using str()
str(feedback)


# TODO: Display first 5 rows of feedback
head(feedback, 5)


cat("\n=== TRANSACTION DATA ===\n")
# TODO: Display structure of transactions
str(transactions)


# TODO: Display first 5 rows of transactions
head(transactions, 5)


cat("\n=== PRODUCT CATALOG DATA ===\n")
# TODO: Display structure of products
str(products)


# TODO: Display first 5 rows of products
head(products, 5)



=== CUSTOMER FEEDBACK DATA ===
spc_tbl_ [100 × 6] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ FeedbackID   : num [1:100] 1 2 3 4 5 6 7 8 9 10 ...
 $ CustomerID   : num [1:100] 12 40 34 1 47 13 13 37 49 23 ...
 $ Customer_Name: chr [1:100] "Bob Wilson" "sarah johnson" "jane smith" "JANE SMITH" ...
 $ Feedback_Text: chr [1:100] "Highly recommend this item" "Excellent service" "Poor quality control" "average product, nothing special" ...
 $ Contact_Info : chr [1:100] "bob.wilson@test.org" "555-123-4567" "jane_smith@company.com" "jane_smith@company.com" ...
 $ Feedback_Date: Date[1:100], format: "2024-02-23" "2024-01-21" ...
 - attr(*, "spec")=
  .. cols(
  ..   FeedbackID = col_double(),
  ..   CustomerID = col_double(),
  ..   Customer_Name = col_character(),
  ..   Feedback_Text = col_character(),
  ..   Contact_Info = col_character(),
  ..   Feedback_Date = col_date(format = "")
  .. )
 - attr(*, "problems")=<externalptr> 


FeedbackID CustomerID Customer_Name Feedback_Text                   
1 1          12         Bob Wilson    Highly recommend this item      
2 2          40         sarah johnson Excellent service               
3 3          34         jane smith    Poor quality control            
4 4           1         JANE SMITH    average product, nothing special
5 5          47         michael brown AMAZING customer support!!!     
  Contact_Info           Feedback_Date
1 bob.wilson@test.org    2024-02-23   
2 555-123-4567           2024-01-21   
3 jane_smith@company.com 2023-09-02   
4 jane_smith@company.com 2023-08-21   
5 555-123-4567           2023-04-24


=== TRANSACTION DATA ===
'data.frame':	150 obs. of  5 variables:
 $ LogID               : int  1 2 3 4 5 6 7 8 9 10 ...
 $ CustomerID          : int  26 21 12 6 32 27 31 30 31 13 ...
 $ Transaction_DateTime: chr  "4/5/24 14:30" "3/15/24 14:30" "3/15/24 14:30" "3/20/24 9:15" ...
 $ Amount              : num  277 175 252 215 269 ...
 $ Status              : chr  "Pending" "Pending" "Pending" "Pending" ...


LogID CustomerID Transaction_DateTime Amount Status   
1 1     26         4/5/24 14:30         277.22 Pending  
2 2     21         3/15/24 14:30        175.16 Pending  
3 3     12         3/15/24 14:30        251.71 Pending  
4 4      6         3/20/24 9:15         214.98 Pending  
5 5     32         3/20/24 9:15         268.91 Completed


=== PRODUCT CATALOG DATA ===
spc_tbl_ [75 × 5] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ ProductID          : num [1:75] 1 2 3 4 5 6 7 8 9 10 ...
 $ Product_Description: chr [1:75] "Apple iPhone 14 Pro - 128GB - Space Black" "samsung galaxy s23 ultra 256gb" "Apple iPhone 14 Pro - 128GB - Space Black" "Apple iPhone 14 Pro - 128GB - Space Black" ...
 $ Category           : chr [1:75] "TV" "TV" "Audio" "Shoes" ...
 $ Price              : num [1:75] 964 1817 853 649 586 ...
 $ In_Stock           : chr [1:75] "Limited" "Yes" "Yes" "Yes" ...
 - attr(*, "spec")=
  .. cols(
  ..   ProductID = col_double(),
  ..   Product_Description = col_character(),
  ..   Category = col_character(),
  ..   Price = col_double(),
  ..   In_Stock = col_character()
  .. )
 - attr(*, "problems")=<externalptr> 


ProductID Product_Description                       Category    Price  
1 1         Apple iPhone 14 Pro - 128GB - Space Black TV           963.53
2 2         samsung galaxy s23 ultra 256gb            TV          1817.44
3 3         Apple iPhone 14 Pro - 128GB - Space Black Audio        852.79
4 4         Apple iPhone 14 Pro - 128GB - Space Black Shoes        648.58
5 5         samsung galaxy s23 ultra 256gb            Electronics  586.35
  In_Stock
1 Limited 
2 Yes     
3 Yes     
4 Yes     
5 Limited

## Part 2: String Cleaning and Standardization

**Business Context:** Product names and feedback text often have inconsistent formatting that prevents accurate analysis.

**Your Tasks:**
1. Clean product names (remove extra spaces, standardize case)
2. Standardize product categories
3. Clean customer feedback text
4. Extract customer names from feedback

**Key Functions:** `str_trim()`, `str_squish()`, `str_to_lower()`, `str_to_upper()`, `str_to_title()`

In [53]:
# Task 2.1: Clean Product Names
# TODO: Create a new column 'product_name_clean' that:
#   - Removes leading/trailing whitespace using str_trim()
#   - Converts to Title Case using str_to_title()
library(dplyr)
library(stringr)

products_clean <- products %>%
  mutate(
    product_name_clean = str_to_title(str_trim(Product_Description))
    # Your code here:
  )

# Display before and after
cat("Product Name Cleaning Results:\n")
products_clean %>%
  select(Product_Description, product_name_clean) %>%
  head(10) %>%
  print()

Product Name Cleaning Results:
# A tibble: 10 × 2
   Product_Description                         product_name_clean               
   <chr>                                       <chr>                            
 1 "Apple iPhone 14 Pro - 128GB - Space Black" "Apple Iphone 14 Pro - 128gb - S…
 2 "samsung galaxy s23 ultra 256gb"            "Samsung Galaxy S23 Ultra 256gb" 
 3 "Apple iPhone 14 Pro - 128GB - Space Black" "Apple Iphone 14 Pro - 128gb - S…
 4 "Apple iPhone 14 Pro - 128GB - Space Black" "Apple Iphone 14 Pro - 128gb - S…
 5 "samsung galaxy s23 ultra 256gb"            "Samsung Galaxy S23 Ultra 256gb" 
 6 "Apple iPhone 14 Pro - 128GB - Space Black" "Apple Iphone 14 Pro - 128gb - S…
 7 "DELL XPS 13 Laptop - Intel i7 - 16GB RAM"  "Dell Xps 13 Laptop - Intel I7 -…
 8 "hp envy printer - wireless - color"        "Hp Envy Printer - Wireless - Co…
 9 "Nike Air Max 270 - Size 10 - Black/White"  "Nike Air Max 270 - Size 10 - Bl…
10 "LG 55\" 4K Smart TV - OLED Display"        "Lg 55\" 4k 

In [54]:
# Task 2.2: Standardize Product Categories
# TODO: Create a new column 'category_clean' that:
#   - Converts category to Title Case
#   - Removes any extra whitespace
library(dplyr)
library(stringr)

products_clean <- products_clean %>%
  mutate(
    category_clean = str_to_title(str_trim(Category)) 
    # Your code here:
    
  )

# Show unique categories before and after
cat("Original categories:\n")
print(unique(products$Category))

cat("\nCleaned categories:\n")
print(unique(products_clean$category_clean))

Original categories:
[1] "TV"          "Audio"       "Shoes"       "Electronics" "Computers"  

Cleaned categories:
[1] "Tv"          "Audio"       "Shoes"       "Electronics" "Computers"  


In [55]:
# Task 2.3: Clean Customer Feedback Text
# TODO: Create a new column 'feedback_clean' that:
#   - Converts text to lowercase using str_to_lower()
#   - Removes extra whitespace using str_squish()
library(dplyr)
library(stringr)

feedback_clean <- feedback %>%
  mutate(
    feedback_clean = str_squish(str_to_lower(Feedback_Text))
    # Your code here:
    
  )

# Display sample
cat("Feedback Cleaning Sample:\n")
feedback_clean %>%
  select(Feedback_Text, feedback_clean) %>%
  head(5) %>%
  print()

Feedback Cleaning Sample:
# A tibble: 5 × 2
  Feedback_Text                    feedback_clean                  
  <chr>                            <chr>                           
1 Highly recommend this item       highly recommend this item      
2 Excellent service                excellent service               
3 Poor quality control             poor quality control            
4 average product, nothing special average product, nothing special
5 AMAZING customer support!!!      amazing customer support!!!     


## Part 3: Pattern Detection and Extraction

**Business Context:** Identifying products with specific features and extracting specifications helps with inventory management and marketing.

**Your Tasks:**
1. Identify products with specific keywords (wireless, premium, gaming)
2. Extract numerical specifications from product names
3. Detect sentiment words in customer feedback
4. Extract email addresses from feedback

**Key Functions:** `str_detect()`, `str_extract()`, `str_count()`

In [56]:
# Task 3.1: Detect Product Features
# TODO: Create three new columns:
#   - is_wireless: TRUE if product name contains "wireless" (case-insensitive)
#   - is_premium: TRUE if product name contains "pro", "premium", or "deluxe"
#   - is_gaming: TRUE if product name contains "gaming" or "gamer"
# Hint: Use str_detect() with str_to_lower() for case-insensitive matching
# Hint: Use | (pipe) in regex for OR conditions
library(dplyr)
library(stringr)

products_clean <- products_clean %>%
  mutate(
    is_wireless = str_detect(str_to_lower(product_name_clean), "wireless"),
    is_premium = str_detect(str_to_lower(product_name_clean), "pro|premium|deluxe"),
    is_gaming = str_detect(str_to_lower(product_name_clean), "gaming|
    # Your code here:
gamer")
    
  )

# Display results
cat ("Product Feature Detection:\n") 
products_clean %>%
  select(product_name_clean, is_wireless, is_premium, is_gaming) %>%
  head(10) %>%
  print()

# Summary statistics
cat("\nFeature Summary:\n")
cat("Wireless products:", sum(products_clean$is_wireless), "\n")
cat("Premium products:", sum(products_clean$is_premium), "\n")
cat("Gaming products:", sum(products_clean$is_gaming), "\n")

Product Feature Detection:
# A tibble: 10 × 4
   product_name_clean                          is_wireless is_premium is_gaming
   <chr>                                       <lgl>       <lgl>      <lgl>    
 1 "Apple Iphone 14 Pro - 128gb - Space Black" FALSE       TRUE       FALSE    
 2 "Samsung Galaxy S23 Ultra 256gb"            FALSE       FALSE      FALSE    
 3 "Apple Iphone 14 Pro - 128gb - Space Black" FALSE       TRUE       FALSE    
 4 "Apple Iphone 14 Pro - 128gb - Space Black" FALSE       TRUE       FALSE    
 5 "Samsung Galaxy S23 Ultra 256gb"            FALSE       FALSE      FALSE    
 6 "Apple Iphone 14 Pro - 128gb - Space Black" FALSE       TRUE       FALSE    
 7 "Dell Xps 13 Laptop - Intel I7 - 16gb Ram"  FALSE       FALSE      FALSE    
 8 "Hp Envy Printer - Wireless - Color"        TRUE        FALSE      FALSE    
 9 "Nike Air Max 270 - Size 10 - Black/White"  FALSE       FALSE      FALSE    
10 "Lg 55\" 4k Smart Tv - Oled Display"        FALSE       FALSE      FALS

In [57]:
# Task 3.2: Extract Product Specifications
# TODO: Create a new column 'size_number' that extracts the first number from product_name
# Hint: Use str_extract() with pattern "\\\\d+" to match one or more digits

products_clean <- products_clean %>%
  mutate(
    size_number = str_extract(product_name_clean, "\\d+")
    # Your code here:
  
  )

# Display products with extracted sizes
cat("Extracted Product Specifications:\n")
products_clean %>%
  filter(!is.na(size_number)) %>%
  select(product_name_clean, size_number) %>%
  head(10) %>%
  print()

Extracted Product Specifications:
# A tibble: 10 × 2
   product_name_clean                          size_number
   <chr>                                       <chr>      
 1 "Apple Iphone 14 Pro - 128gb - Space Black" 14         
 2 "Samsung Galaxy S23 Ultra 256gb"            23         
 3 "Apple Iphone 14 Pro - 128gb - Space Black" 14         
 4 "Apple Iphone 14 Pro - 128gb - Space Black" 14         
 5 "Samsung Galaxy S23 Ultra 256gb"            23         
 6 "Apple Iphone 14 Pro - 128gb - Space Black" 14         
 7 "Dell Xps 13 Laptop - Intel I7 - 16gb Ram"  13         
 8 "Nike Air Max 270 - Size 10 - Black/White"  270        
 9 "Lg 55\" 4k Smart Tv - Oled Display"        55         
10 "Sony Wh-1000xm4 Wireless Headphones"       1000       


In [58]:
# Task 3.3: Simple Sentiment Analysis
# TODO: Create three new columns:
#   - positive_words: count of positive words ("great", "excellent", "love", "amazing")
#   - negative_words: count of negative words ("bad", "terrible", "hate", "awful")
#   - sentiment_score: positive_words - negative_words
# Hint: Use str_count() to count pattern occurrences

feedback_clean <- feedback_clean %>%
  mutate(
    positive_words = str_count(feedback_clean, "great|excellent|love|amazing"),
    negative_words = str_count(feedback_clean, "bad|terrible|hate|awful"),
    sentiment_score = positive_words - negative_words
    # Your code here:
    
  )

# Display sentiment analysis results
cat("Sentiment Analysis Results:\n")
feedback_clean %>%
  select(feedback_clean, positive_words, negative_words, sentiment_score) %>%
  head(10) %>%
  print()

# Summary
cat("\nOverall Sentiment Summary:\n")
cat("Average sentiment score:", mean(feedback_clean$sentiment_score), "\n")
cat("Positive reviews:", sum(feedback_clean$sentiment_score > 0), "\n")
cat("Negative reviews:", sum(feedback_clean$sentiment_score < 0), "\n")

Sentiment Analysis Results:
# A tibble: 10 × 4
   feedback_clean                  positive_words negative_words sentiment_score
   <chr>                                    <int>          <int>           <int>
 1 highly recommend this item                   0              0               0
 2 excellent service                            1              0               1
 3 poor quality control                         0              0               0
 4 average product, nothing speci…              0              0               0
 5 amazing customer support!!!                  1              0               1
 6 amazing customer support!!!                  1              0               1
 7 average product, nothing speci…              0              0               0
 8 good value for money                         0              0               0
 9 highly recommend this item                   0              0               0
10 highly recommend this item                   0             

## Part 4: Date Parsing and Component Extraction

**Business Context:** Transaction dates need to be parsed and analyzed to understand customer behavior patterns.

**Your Tasks:**
1. Parse transaction dates from text to Date objects
2. Extract date components (year, month, day, weekday)
3. Identify weekend vs weekday transactions
4. Extract quarter and month names

**Key Functions:** `ymd()`, `mdy()`, `dmy()`, `year()`, `month()`, `day()`, `wday()`, `quarter()`

In [59]:
# Task 4.1: Parse Transaction Dates
# TODO: Create a new column 'date_parsed' that parses the transaction_date column
# Hint: Check the format of transaction_date first, then use ymd(), mdy(), or dmy()

transactions_clean <- transactions %>%
  mutate(
    date_parsed = mdy(Transaction_DateTime)
    # Your code here:
    
  )

# Verify parsing worked
cat("Date Parsing Results:\n")
transactions_clean %>%
  select(Transaction_DateTime, date_parsed) %>%
  head(10) %>%
  print()

Warning message:
“There was 1 warning in `mutate()`.
ℹ In argument: `date_parsed = mdy(Transaction_DateTime)`.
Caused by warning:
! All formats failed to parse. No formats found.”


Date Parsing Results:
   Transaction_DateTime date_parsed
1          4/5/24 14:30        <NA>
2         3/15/24 14:30        <NA>
3         3/15/24 14:30        <NA>
4          3/20/24 9:15        <NA>
5          3/20/24 9:15        <NA>
6          3/20/24 9:15        <NA>
7          3/20/24 9:15        <NA>
8         3/15/24 14:30        <NA>
9   25-03-2024 16:45:30        <NA>
10         4/5/24 14:30        <NA>


In [60]:
# Task 4.2: Extract Date Components
# TODO: Create the following new columns:
#   - trans_year: Extract year from date_parsed
#   - trans_month: Extract month number from date_parsed
#   - trans_month_name: Extract month name (use label=TRUE, abbr=FALSE)
#   - trans_day: Extract day of month from date_parsed
#   - trans_weekday: Extract weekday name (use label=TRUE, abbr=FALSE)
#   - trans_quarter: Extract quarter from date_parsed

transactions_clean <- transactions_clean %>%
  mutate(
    trans_year = year(date_parsed),
    trans_month = month(date_parsed),
    trans_month_name = month(date_parsed, label=TRUE, abbr=FALSE),
    trans_day = day(date_parsed),
    trans_weekday = wday(date_parsed, label=TRUE, abbr=FALSE),  
    trans_quarter = quarter(date_parsed)
    # Your code here:
    
  )

# Display results
cat("Date Component Extraction:\n")
transactions_clean %>%
  select(date_parsed, trans_month_name, trans_weekday, trans_quarter) %>%
  head(10) %>%
  print()

Date Component Extraction:
   date_parsed trans_month_name trans_weekday trans_quarter
1         <NA>             <NA>          <NA>            NA
2         <NA>             <NA>          <NA>            NA
3         <NA>             <NA>          <NA>            NA
4         <NA>             <NA>          <NA>            NA
5         <NA>             <NA>          <NA>            NA
6         <NA>             <NA>          <NA>            NA
7         <NA>             <NA>          <NA>            NA
8         <NA>             <NA>          <NA>            NA
9         <NA>             <NA>          <NA>            NA
10        <NA>             <NA>          <NA>            NA


In [61]:
# Task 4.3: Identify Weekend Transactions
# TODO: Create a new column 'is_weekend' that is TRUE if the transaction was on Saturday or Sunday
# Hint: Use wday() which returns 1 for Sunday and 7 for Saturday
# Hint: Use %in% c(1, 7) to check if day is weekend

transactions_clean <- transactions_clean %>%
  mutate(
    is_weekend = wday(date_parsed) %in% c(1, 7)
    # Your code here:
    
  )

# Summary
cat("Weekend vs Weekday Transactions:\n")
table(transactions_clean$is_weekend) %>% print()

cat("\nPercentage of weekend transactions:",
    round(sum(transactions_clean$is_weekend) / nrow(transactions_clean) * 100, 1), "%\n")

Weekend vs Weekday Transactions:

FALSE 
  150 

Percentage of weekend transactions: 0 %


## Part 5: Date Calculations and Customer Recency Analysis

**Business Context:** Understanding how recently customers transacted helps identify at-risk customers for re-engagement campaigns.

**Your Tasks:**
1. Calculate days since each transaction
2. Categorize customers by recency (Recent, Moderate, Old)
3. Identify customers who haven't transacted in 90+ days
4. Calculate average days between transactions per customer

**Key Functions:** `today()`, date arithmetic, `case_when()`

In [62]:
# Task 5.1: Calculate Days Since Transaction
# TODO: Create a new column 'days_since' that calculates days from date_parsed to today()
# Hint: Use as.numeric(today() - date_parsed)

transactions_clean <- transactions_clean %>%
  mutate(
    days_since = as.numeric(today() - date_parsed)
    # Your code here:
    
  )

# Display results
cat("Days Since Transaction:\n")
transactions_clean %>%
  select(CustomerID, date_parsed, days_since) %>%
  arrange(desc(days_since)) %>%
  head(10) %>%
  print()

Days Since Transaction:
   CustomerID date_parsed days_since
1          26        <NA>         NA
2          21        <NA>         NA
3          12        <NA>         NA
4           6        <NA>         NA
5          32        <NA>         NA
6          27        <NA>         NA
7          31        <NA>         NA
8          30        <NA>         NA
9          31        <NA>         NA
10         13        <NA>         NA


In [63]:
# Task 5.2: Categorize by Recency
# TODO: Create a new column 'recency_category' using case_when():
#   - "Recent" if days_since <= 30
#   - "Moderate" if days_since <= 90
#   - "At Risk" if days_since > 90

transactions_clean <- transactions_clean %>%
  mutate(
    recency_category = case_when(
      days_since <= 30 ~ "Recent",
      days_since <= 90 ~ "Moderate",
      days_since > 90 ~ "At Risk"
    ) 
    # Your code here:
    
  )

# Display distribution
cat("Recency Category Distribution:\n")
table(transactions_clean$recency_category) %>% print()

# Show at-risk customers
cat("\nAt-Risk Customers (>90 days):\n")
transactions_clean %>%
  filter(recency_category == "At Risk") %>%
  select(CustomerID, date_parsed, days_since) %>%
  arrange(desc(days_since)) %>%
  print()

Recency Category Distribution:
< table of extent 0 >

At-Risk Customers (>90 days):
[1] CustomerID  date_parsed days_since 
<0 rows> (or 0-length row.names)


## Part 6: Combined String and Date Operations

**Business Context:** Create personalized customer outreach messages based on purchase recency.

**Your Tasks:**
1. Extract first names from customer names
2. Create personalized messages based on recency
3. Analyze transaction patterns by weekday
4. Identify best customers (recent + high value)

**Key Functions:** Combine `str_extract()`, date calculations, `case_when()`, `group_by()`, `summarize()`

In [64]:
# Task 6.1: Extract First Names and Create Personalized Messages
# TODO: Create two new columns:
#   - first_name: Extract first name from customer_name (everything before first space)
#   - personalized_message: Create message based on recency_category
#     * Recent: "Hi [name]! Thanks for your recent purchase!"
#     * Moderate: "Hi [name], we miss you! Check out our new products."
#     * At Risk: "Hi [name], it's been a while! Here's a special offer for you."
# Hint: Use str_extract() with pattern "^\\\\w+" for first name
# Hint: Use paste() to combine strings in case_when()

customer_outreach <- transactions_clean %>%
  mutate(
    first_name = str_extract(CustomerID, "^\\w+"),
    personalized_message = case_when(
      recency_category == "Recent" ~ paste("Hi", first_name, "! Thanks for your recent purchase!"),
      recency_category == "Moderate" ~ paste("Hi", first_name, ", we miss you! Check out our new products."),
      recency_category == "At Risk" ~ paste("Hi", first_name, ", it's been a while! Here's a special offer for you.")
    )
    # Your code here:
    
  )

# Display personalized messages
cat("Personalized Customer Messages:\n")
customer_outreach %>%
  select(CustomerID, first_name, days_since, personalized_message) %>%
  head(10) %>%
  print()

Personalized Customer Messages:
   CustomerID first_name days_since personalized_message
1          26         26         NA                 <NA>
2          21         21         NA                 <NA>
3          12         12         NA                 <NA>
4           6          6         NA                 <NA>
5          32         32         NA                 <NA>
6          27         27         NA                 <NA>
7          31         31         NA                 <NA>
8          30         30         NA                 <NA>
9          31         31         NA                 <NA>
10         13         13         NA                 <NA>


In [65]:
# Task 6.2: Analyze Transaction Patterns by Weekday
# TODO: Group by trans_weekday and calculate:
#   - transaction_count: number of transactions
#   - total_amount: sum of amount (if available)
#   - avg_amount: average amount per transaction
# TODO: Arrange by transaction_count descending

weekday_patterns <- transactions_clean %>%
  group_by(trans_weekday) %>%
  summarize(
    transaction_count = n(),
    total_amount = sum(Amount, na.rm=TRUE),
    avg_amount = mean(Amount, na.rm=TRUE)
  ) %>%
  arrange(desc(transaction_count))
  # Your code here:
  

# Display results
cat("Transaction Patterns by Weekday:\n")
print(weekday_patterns)

# Identify busiest day
busiest_day <- weekday_patterns$trans_weekday[1]
cat("\n🔥 Busiest day:", as.character(busiest_day), "\n")

Transaction Patterns by Weekday:
# A tibble: 1 × 4
  trans_weekday transaction_count total_amount avg_amount
  <ord>                     <int>        <dbl>      <dbl>
1 NA                          150       37734.       252.

🔥 Busiest day: NA 


In [66]:
# Task 6.3: Monthly Transaction Analysis
# TODO: Group by trans_month_name and calculate:
#   - transaction_count
#   - unique_customers: use n_distinct(customer_name)
# TODO: Arrange by trans_month (to show chronological order)

monthly_patterns <- transactions_clean %>%
  group_by(trans_month, trans_month_name) %>%
  summarize(
    transaction_count = n(),
    unique_customers = n_distinct(CustomerID)
  ) %>%
  arrange(trans_month)
  # Your code here:
  

# Display results
cat("Monthly Transaction Patterns:\n")
print(monthly_patterns)

`summarise()` has grouped output by 'trans_month'. You can override using the
`.groups` argument.


Monthly Transaction Patterns:
# A tibble: 1 × 4
# Groups:   trans_month [1]
  trans_month trans_month_name transaction_count unique_customers
        <dbl> <ord>                        <int>            <int>
1          NA NA                             150               47


## Part 7: Business Intelligence Summary

**Business Context:** Create an executive summary that combines all your analyses into actionable insights.

**Your Tasks:**
1. Calculate key metrics across all datasets
2. Identify top products and categories
3. Summarize customer sentiment
4. Provide data-driven recommendations

In [67]:
# Task 7.1: Create Business Intelligence Dashboard

cat("\n", rep("=", 60), "\n")
cat("         BUSINESS INTELLIGENCE SUMMARY\n")
cat(rep("=", 60), "\n\n")

# Product Analysis
cat("📦 PRODUCT ANALYSIS\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Total number of products
#   - Number of wireless products
#   - Number of premium products
#   - Most common category
total_products <- nrow(products_clean)
wireless_products <- sum(products_clean$is_wireless)
premium_products <- sum(products_clean$is_premium)
most_common_category <- products_clean %>%
count(Category) %>%
arrange(desc(n)) %>%
slice(1) %>%
pull(Category)


# Customer Sentiment
cat("\n💬 CUSTOMER SENTIMENT\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Total feedback entries
#   - Average sentiment score
#   - Percentage of positive reviews
#   - Percentage of negative reviews
total_feedback <- nrow(feedback_clean)
avg_sentiment <- mean(feedback_clean$sentiment_score)
positive_reviews_pct <- mean(feedback_clean$sentiment_score > 0) * 100
negative_reviews_pct <- mean(feedback_clean$sentiment_score < 0) * 100


# Transaction Patterns
cat("\n📊 TRANSACTION PATTERNS\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Total transactions
#   - Date range (earliest to latest)
#   - Busiest weekday
#   - Weekend transaction percentage
total_transactions <- nrow(transactions_clean)
date_range_start <- min(transactions_clean$date_parsed, na.rm=TRUE)
date_range_end <- max(transactions_clean$date_parsed, na.rm=TRUE)
busiest_weekday <- weekday_patterns$trans_weekday[which.max(weekday_patterns$transaction_count)]
weekend_transaction_pct <- round(sum(transactions_clean$is_weekend) / total_transactions * 100, 1)


# Customer Recency
cat("\n👥 CUSTOMER RECENCY\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Number of recent customers (< 30 days)
#   - Number of at-risk customers (> 90 days)
#   - Percentage needing re-engagement
recent_customers <- sum(transactions_clean$recency_category == "Recent")
at_risk_customers <- sum(transactions_clean$recency_category == "At Risk")
percentage_reengagement <- round(at_risk_customers / nrow(transactions_clean) * 100, 1)



 = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = 
         BUSINESS INTELLIGENCE SUMMARY
= = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = 

📦 PRODUCT ANALYSIS
─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ 

💬 CUSTOMER SENTIMENT
─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ 

📊 TRANSACTION PATTERNS
─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ 


Warning message in min.default(structure(c(NA_real_, NA_real_, NA_real_, NA_real_, :
“no non-missing arguments to min; returning Inf”
Warning message in max.default(structure(c(NA_real_, NA_real_, NA_real_, NA_real_, :
“no non-missing arguments to max; returning -Inf”



👥 CUSTOMER RECENCY
─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ 


In [68]:
# Task 7.2: Identify Top Products by Category
# TODO: Group products by category_clean and count products in each
# TODO: Arrange by count descending
# TODO: Display top 5 categories

top_categories <- products_clean %>%
  group_by(Category) %>%
  summarize(product_count = n()) %>%
  arrange(desc(product_count)) %>%
  slice_head(n = 5)
  # Your code here:
  

cat("Top Product Categories:\n")
print(top_categories)

Top Product Categories:
# A tibble: 5 × 2
  Category    product_count
  <chr>               <int>
1 Electronics            21
2 Computers              15
3 Audio                  14
4 TV                     14
5 Shoes                  11


## Part 8: Reflection Questions

Answer the following questions based on your analysis. Write your answers in the markdown cells below.

### Question 8.1: Data Quality Impact

**How did cleaning the text data (removing spaces, standardizing case) improve your ability to analyze the data? Provide specific examples from your homework.**

Your answer here: Cleaning the text data significantly improved my ability to categorize and analyze the data by ensuring consistency. For example, standardizing product names to title case allowed me to accurately group products into categories without misclassifications due to formatting inconsistencies. This enabled me to identify the most popular product categories effectively.



### Question 8.2: Pattern Detection Value

**What business insights did you gain from detecting patterns in product names (wireless, premium, gaming)? How could a business use this information?**

Your answer here: By detetcting patterns in product names such as wireless, premium, and gaming, I was able to identify which products had specific features that were likely to appeal to certain customer segments. For example, wireless products may attract tech-savvy customers who prioritize convenience, while premium products may appeal to customers looking for high-quality or luxury items. A business could use this information to tailor their marketing strategies, create targeted promotions, and optimize inventory management by stocking more of the popular product types that align with customer preferences.



### Question 8.3: Date Analysis Importance

**Why is analyzing transaction dates by weekday and month important for business operations? Provide at least three specific business applications.**

Your answer here:
Analyzing transaction dates by weekday and month is important for business operations because it helps in:
1. Analyzing and understanding sales trends which helps in knowing your inventory.
2. Understanding what marketing campaigns are effective and what is popular among customers.
3. Understanding what customers are purchasing which then can be used to improve customer experience and have those products readily available.


### Question 8.4: Customer Recency Strategy

**Based on your recency analysis, what specific actions would you recommend for customers in each category (Recent, Moderate, At Risk)? How would you prioritize these actions?**

Your answer here: For customers in the recent category, I would recommend sending frequent emails and offering rewards so they can stay and be engaged. For customers in the moderate category, I would give them more targeted promotions or discounts so they can be motivated to stay and be there. For customers in the at-risk category, I would include them in special re-engagement campaigns with exclusive offers to bring them back.
The priority would be highest for the at-risk customers, followed by moderate, and then recent customers.



### Question 8.5: Sentiment Analysis Application

**How could the sentiment analysis you performed be used to improve products or customer service? What are the limitations of this simple sentiment analysis approach?**

Your answer here: The sentiment analysis could be used to improve products or customer service by identifying common themes in negative feedback and addressing those issues directly. For example, if many customers mention poor battery life in their reviews, the company could focus on improving that aspect in future product iterations. Additionally, positive feedback can highlight what customers appreciate, allowing the business to reinforce those strengths in marketing and product development. The limitations of this simple sentiment analysis approach include its reliance on keyword detection, which may not capture the full context of customer sentiment. It may miss nuances such as sarcasm or mixed feelings, and it may not account for the intensity of sentiment expressed in the feedback.



### Question 8.6: Real-World Application

**Describe a real business scenario where you would need to combine string manipulation and date analysis (like you did in this homework). What insights would you be trying to discover?**

Your answer here: A real business scenario where I would need to combine string manipulation and date analysis is in retail marketing would be for a beauty company like Sephora. By analyzing customer reviews alongside their purchase dates the company could identify trends in product popularity over time. For example, they could discover that certain skincare products receive more positive feedback during specific seasons or months, allowing them to tailor marketing campaigns and inventory management accordingly. Additionally, they could track how customer sentiment changes over time after product launches or promotional events, providing insights into customer satisfaction and loyalty.



## Summary and Submission

### What You've Accomplished

In this homework, you've successfully:
- ✅ Cleaned and standardized messy text data using `stringr` functions
- ✅ Detected patterns and extracted information from text
- ✅ Parsed dates and extracted temporal components using `lubridate`
- ✅ Calculated customer recency for segmentation
- ✅ Analyzed transaction patterns by time periods
- ✅ Combined string and date operations for business insights
- ✅ Created personalized customer communications
- ✅ Generated executive-ready business intelligence summaries

### Key Skills Mastered

**String Manipulation:**
- `str_trim()`, `str_squish()` - Whitespace handling
- `str_to_lower()`, `str_to_upper()`, `str_to_title()` - Case conversion
- `str_detect()` - Pattern detection
- `str_extract()` - Information extraction
- `str_count()` - Pattern counting

**Date/Time Operations:**
- `ymd()`, `mdy()`, `dmy()` - Date parsing
- `year()`, `month()`, `day()`, `wday()` - Component extraction
- `quarter()` - Period extraction
- `today()` - Current date
- Date arithmetic - Calculating differences

**Business Applications:**
- Data cleaning and standardization
- Customer segmentation by recency
- Sentiment analysis
- Pattern identification
- Temporal trend analysis
- Personalized communication

### Submission Checklist

Before submitting, ensure you have:
- [ ] Entered your name, student ID, and date at the top
- [ ] Completed all code tasks (Parts 1-7)
- [ ] Run all cells successfully without errors
- [ ] Answered all reflection questions (Part 8)
- [ ] Used proper commenting in your code
- [ ] Used the pipe operator (`%>%`) where appropriate
- [ ] Verified your results make business sense
- [ ] Checked for any remaining TODO comments

### Grading Criteria

Your homework will be evaluated on:
- **Code Correctness (40%)**: All tasks completed correctly
- **Code Quality (20%)**: Clean, well-commented, efficient code
- **Business Understanding (20%)**: Demonstrates understanding of business applications
- **Reflection Questions (15%)**: Thoughtful, complete answers
- **Presentation (5%)**: Professional formatting and organization

### Next Steps

In Lesson 8, you'll learn:
- Advanced data wrangling with complex pipelines
- Sophisticated conditional logic with `case_when()`
- Data validation and quality checks
- Creating reproducible analysis workflows
- Professional best practices for business analytics

**Great work on completing this assignment! 🎉**